tap_pendo/streams.py

```diff
@@ -416,22 +416,28 @@ def send_request_get_results(self, req):
         # If value is 0,"0", "" or None then it will set default to default to 300.0 seconds if not passed in config.
         config_request_timeout = self.config.get('request_timeout')
         request_timeout = config_request_timeout and float(config_request_timeout) or REQUEST_TIMEOUT # pylint: disable=consider-using-ternary
-        with session.send(req, stream=True, timeout=request_timeout) as resp:
-            if 'Too Many Requests' in resp.reason:
-                retry_after = 30
-                LOGGER.info("Rate limit reached. Sleeping for %s seconds",
-                            retry_after)
-                time.sleep(retry_after)
-                raise Server42xRateLimitError(resp.reason)
-
-            resp.raise_for_status() # Check for requests status and raise exception in failure
-
-            # Return list of records instead of yielding because more than one iteration occur over data in tap flow
-            # and yield will return generator which flushes out after one iteration.
-            to_return = []
-            for item in ijson.items(resp.raw, 'results.item'):
-                to_return.append(humps.decamelize(item))
-            return to_return
+        # Send request for stream data directly(without 'with' statement) so that
+        # exception handling and yielding can be utilized properly below
+        resp = session.send(req, stream=True, timeout=request_timeout)
+
+        if 'Too Many Requests' in resp.reason:
+            retry_after = 30
+            LOGGER.info("Rate limit reached. Sleeping for %s seconds",
+                        retry_after)
+            time.sleep(retry_after)
+            raise Server42xRateLimitError(resp.reason)
+
+        resp.raise_for_status() # Check for requests status and raise exception in failure
+
+        # Separated yielding of records into a new function and called that here
+        # so that any exception raised from the session.send can be thrown from here and
+        # handled properly using backoff on request function.
+        return self.send_records(resp)
```

Comment:
I think the proper way to accomplish this would be `yield from self.send_records()`, though I don't know if that changes how the exceptions would be thrown.

In this function call, I would expect any exception thrown to be a snapped connection VERY far into the sync (6 to 10 hours or so). In that case, I'm not sure that retrying the same request is the right path, since from the user's perspective, it would seem that the tap is just syncing way more records than exist and never finishing.

For that sort of scenario, I would hope that we can rely on sorting in the initial request and save bookmarks to resume the next time the tap runs.

Reply:
Can you point me to a test that shows the Exception handling?

Reply:
Here are some of the unit tests for the timeout, connection errors.
https://github.com/singer-io/tap-pendo/blob/3e69f33bf9839957fb31fa2d5e96d34b9e0531a6/tests/unittests/test_timeout_value.py#L132

https://github.com/singer-io/tap-pendo/blob/3e69f33bf9839957fb31fa2d5e96d34b9e0531a6/tests/unittests/test_backoff_for_conn_reset.py#L15

https://github.com/singer-io/tap-pendo/blob/3e69f33bf9839957fb31fa2d5e96d34b9e0531a6/tests/unittests/test_backoff.py#L234

Reply:
@savan-chovatiya I see. I was concerned about how long it takes for `ConnectionResetByPeer` to be thrown, but looking at our historical data over the past month, I only see a max time of ~19 minutes, so the retries seem good here.